Including `finish-chunk` in `test-aligns-temporal-columns-correctly-363` breaks the test

Stacktrace

```
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1221)
	at xtdb.indexer.live_index.LiveIndex.liveTable(live_index.clj:216)
	at xtdb.indexer.live_index.LiveIndex$reify$reify__19055.apply(live_index.clj:228)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at xtdb.indexer.live_index.LiveIndex$reify__19054.liveTable(live_index.clj:225)
	at xtdb.indexer$__GT_put_indexer$fn__23590.invoke(indexer.clj:81)
	at clojure.core$mapv$fn__8551.invoke(core.clj:6988)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:49)
	at clojure.core.protocols$fn__8246.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8246.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8194$G__8189__8207.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6895)
	at clojure.core$mapv.invokeStatic(core.clj:6979)
	at clojure.core$mapv.invoke(core.clj:6979)
	at xtdb.indexer$__GT_put_indexer.invokeStatic(indexer.clj:75)
	at xtdb.indexer$__GT_put_indexer.invoke(indexer.clj:67)
	at xtdb.indexer.Indexer$fn__23720$index_tx_ops__23723$fn__23724.invoke(indexer.clj:496)
	at clojure.lang.Delay.deref(Delay.java:42)
	at clojure.core$deref.invokeStatic(core.clj:2337)
	at clojure.core$deref.invoke(core.clj:2323)
	at xtdb.indexer.Indexer$fn__23720$index_tx_ops__23723.invoke(indexer.clj:504)
	at xtdb.indexer.Indexer$fn__23720$fn__23740.invoke(indexer.clj:514)
	at xtdb.indexer.Indexer$fn__23720.invoke(indexer.clj:513)
	at xtdb.indexer.Indexer.indexTx(indexer.clj:479)
```

My suspicion here is that, at least in test, the `tables` HashMap has the potential to be concurrently modified.

In the running system finish chunk and tx-op processing happens within the same loop. In test, we call finish chunk and rely on await-tx to avoid concurrent modification, which should and on the most part does work, but I could imagine it failing.

Speculation here as I was unable to repro, one to watch I think.

--- a/core/src/main/clojure/xtdb/indexer.clj
+++ b/core/src/main/clojure/xtdb/indexer.clj
@@ -468,7 +468,7 @@
                      (pr-str tx-key)
                      (pr-str latest-completed-tx))
 
-          (with-open [live-idx-tx (.startTx live-idx tx-key)]
+          (let [live-idx-tx (.startTx live-idx tx-key)]
             (add-tx-row! row-counter live-idx-tx tx-key
                          (err/illegal-arg :invalid-system-time
                                           {::err/message "specified system-time older than current tx"
@@ -526,9 +526,10 @@
                   (if e
                     (do
                       (when (not= e abort-exn)
-                        (log/debug e "aborted tx"))
+                        (log/debug e "aborted tx")
+                        (.abort live-idx-tx))
 
-                      (with-open [live-idx-tx (.startTx live-idx tx-key)]
+                      (let [live-idx-tx (.startTx live-idx tx-key)]
                         (add-tx-row! row-counter live-idx-tx tx-key e)
                         (.commit live-idx-tx)))
 

--- a/core/src/main/clojure/xtdb/indexer/live_index.clj
+++ b/core/src/main/clojure/xtdb/indexer/live_index.clj
@@ -18,12 +18,12 @@
            (xtdb.trie LiveHashTrie)
            (xtdb.vector IRelationWriter IVectorWriter RelationReader)))
 
-;;
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveTableWatermark
   (^java.util.Map columnTypes [])
   (^xtdb.vector.RelationReader liveRelation [])
-  (^xtdb.trie.LiveHashTrie liveTrie []))
+  (^xtdb.trie.LiveHashTrie liveTrie [])
+  (^void close []))
 
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveTableTx
@@ -32,12 +32,13 @@
   (^void logPut [^java.nio.ByteBuffer iid, ^long validFrom, ^long validTo, writeDocFn])
   (^void logDelete [^java.nio.ByteBuffer iid, ^long validFrom, ^long validTo])
   (^void logEvict [^java.nio.ByteBuffer iid])
-  (^void commit [])
-  (^void close []))
+  (^xtdb.indexer.live_index.ILiveTable commit [])
+  (^void abort []))
 
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveTable
-  (^xtdb.indexer.live_index.ILiveTableTx startTx [^xtdb.api.protocols.TransactionInstant txKey])
+  (^xtdb.indexer.live_index.ILiveTableTx startTx [^xtdb.api.protocols.TransactionInstant txKey
+                                                  ^boolean newLiveTable])
   (^xtdb.indexer.live_index.ILiveTableWatermark openWatermark [^boolean retain])
   (^java.util.concurrent.CompletableFuture #_<List<Map$Entry>> finishChunk [^long chunkIdx])
   (^void close []))
@@ -45,14 +46,15 @@
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveIndexWatermark
   (^java.util.Map allColumnTypes [])
-  (^xtdb.indexer.live_index.ILiveTableWatermark liveTable [^String tableName]))
+  (^xtdb.indexer.live_index.ILiveTableWatermark liveTable [^String tableName])
+  (^void close []))
 
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveIndexTx
   (^xtdb.indexer.live_index.ILiveTableTx liveTable [^String tableName])
   (^xtdb.indexer.live_index.ILiveIndexWatermark openWatermark [])
   (^void commit [])
-  (^void close []))
+  (^void abort []))
 
 #_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}
 (definterface ILiveIndex
@@ -96,7 +98,7 @@
                     ^IVectorWriter delete-wtr, ^IVectorWriter delete-valid-from-wtr, ^IVectorWriter delete-valid-to-wtr
                     ^IVectorWriter evict-wtr]
   ILiveTable
-  (startTx [this-table tx-key]
+  (startTx [this-table tx-key new-live-table?]
     (let [!transient-trie (atom live-trie)
           system-from-µs (util/instant->micros (.system-time tx-key))]
       (reify ILiveTableTx
@@ -157,10 +159,12 @@
 
         (commit [_]
           (locking this-table
-            (set! (.-live-trie this-table) @!transient-trie)))
+            (set! (.-live-trie this-table) @!transient-trie)
+            this-table))
 
-        AutoCloseable
-        (close [_]))))
+        (abort [_]
+          (when new-live-table?
+            (util/close this-table))))))
 
   (finishChunk [_ chunk-idx]
     (let [live-rel-rdr (vw/rel-wtr->rdr live-rel)]
@@ -199,12 +203,12 @@
       (util/close live-rel))))
 
 (defn ->live-table
-  ([allocator object-store table-name] (->live-table allocator object-store table-name {}))
+  (^xtdb.indexer.live_index.ILiveTable [allocator object-store table-name] (->live-table allocator object-store table-name {}))
 
-  ([allocator object-store table-name
-    {:keys [->live-trie]
-     :or {->live-trie (fn [iid-rdr]
-                        (LiveHashTrie/emptyTrie iid-rdr))}}]
+  (^xtdb.indexer.live_index.ILiveTable [allocator object-store table-name
+                                        {:keys [->live-trie]
+                                         :or {->live-trie (fn [iid-rdr]
+                                                            (LiveHashTrie/emptyTrie iid-rdr))}}]
 
    (util/with-close-on-catch [rel (trie/open-leaf-root allocator)]
      (let [iid-wtr (.writerForName rel "xt$iid")
@@ -227,31 +231,35 @@
 
 (defrecord LiveIndex [^BufferAllocator allocator, ^ObjectStore object-store, ^Map tables, ^long log-limit, ^long page-limit]
   ILiveIndex
-  (liveTable [_ table-name]
-    (.computeIfAbsent tables table-name
-                      (reify Function
-                        (apply [_ table-name]
-                          (->live-table allocator object-store table-name
-                                        {:->live-trie (partial ->live-trie log-limit page-limit)})))))
+  (liveTable [_ table-name] (.get tables table-name))
 
-  (startTx [live-idx tx-key]
+  (startTx [this-table tx-key]
     (let [table-txs (HashMap.)]
       (reify ILiveIndexTx
         (liveTable [_ table-name]
           (.computeIfAbsent table-txs table-name
                             (reify Function
                               (apply [_ table-name]
-                                (-> (.liveTable live-idx table-name)
-                                    (.startTx tx-key))))))
+                                (let [live-table (.liveTable this-table table-name)
+                                      new-live-table? (nil? live-table)
+                                      ^ILiveTable live-table (or live-table
+                                                                 (->live-table allocator object-store table-name
+                                                                               {:->live-trie (partial ->live-trie log-limit page-limit)}))]
+
+                                  (.startTx live-table tx-key new-live-table?))))))
 
         (commit [_]
-          (doseq [^ILiveTableTx table-tx (.values table-txs)]
-            (.commit table-tx)))
+          (doseq [[table-name ^ILiveTableTx live-table-tx] table-txs]
+            (.put tables table-name (.commit live-table-tx))))
+
+        (abort [_]
+          (doseq [^ILiveTableTx live-table-tx (.values table-txs)]
+            (.abort live-table-tx)))
 
         (openWatermark [_]
           (util/with-close-on-catch [wms (HashMap.)]
-            (doseq [[table-name ^ILiveTableTx live-table] table-txs]
-              (.put wms table-name (.openWatermark live-table)))
+            (doseq [[table-name ^ILiveTableTx live-table-tx] table-txs]
+              (.put wms table-name (.openWatermark live-table-tx)))
 
             (doseq [[table-name ^ILiveTable live-table] tables]
               (.computeIfAbsent wms table-name
@@ -265,8 +273,7 @@
               (close [_] (util/close wms)))))
 
         AutoCloseable
-        (close [_]
-          (util/close table-txs)))))
+        (close [_]))))
 
   (openWatermark [_]
 

--- a/src/test/clojure/xtdb/indexer/live_index_test.clj
+++ b/src/test/clojure/xtdb/indexer/live_index_test.clj
@@ -38,8 +38,8 @@
                        (mapv (comp vec util/uuid->bytes)))]
 
     (t/testing "commit"
-      (util/with-open [live-idx-tx (.startTx live-index (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))
-                       live-table-tx (.liveTable live-idx-tx "my-table")]
+      (let [live-idx-tx (.startTx live-index (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))
+            live-table-tx (.liveTable live-idx-tx "my-table")]
         (let [wp (IVectorPosition/build)]
           (doseq [^UUID iid iids]
             (.logPut live-table-tx (ByteBuffer/wrap (util/uuid->bytes iid)) 0 0

--- a/src/test/clojure/xtdb/indexer/live_table_test.clj
+++ b/src/test/clojure/xtdb/indexer/live_table_test.clj
@@ -8,15 +8,12 @@
             [xtdb.test-util :as tu]
             [xtdb.util :as util]
             [xtdb.vector.reader :as vr])
-  (:import (java.lang AutoCloseable)
-           (java.nio ByteBuffer)
+  (:import (java.nio ByteBuffer)
            (java.util Arrays HashMap)
            (org.apache.arrow.memory RootAllocator)
-           (xtdb.indexer.live_index
-             ILiveTable ILiveTableTx TestLiveTable
-             ILiveIndex ILiveTableWatermark ILiveIndexWatermark)
-           xtdb.vector.IVectorPosition
-           (xtdb.trie LiveHashTrie LiveHashTrie$Leaf)))
+           (xtdb.indexer.live_index ILiveIndex ILiveTable ILiveTableWatermark TestLiveTable)
+           (xtdb.trie LiveHashTrie LiveHashTrie$Leaf)
+           xtdb.vector.IVectorPosition))
 
 (defn uuid-equal-to-path? [uuid path]
   (Arrays/equals
@@ -33,80 +30,73 @@
       (tu/with-tmp-dirs #{path}
         (with-open [obj-store (obj-store-test/fs path)
                     allocator (RootAllocator.)
-                    live-table ^ILiveTable (live-index/->live-table
-                                            allocator obj-store "foo" {:->live-trie (partial live-index/->live-trie 2 4)})
-                    live-table-tx (.startTx
-                                   live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))]
+                    live-table (live-index/->live-table allocator obj-store "foo" {:->live-trie (partial live-index/->live-trie 2 4)})]
 
-          (let [wp (IVectorPosition/build)]
-            (dotimes [_n n]
-              (.logPut
-               live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
-               0 0 #(.getPositionAndIncrement wp))))
+          (let [live-table-tx (.startTx live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")) false)]
+            (let [wp (IVectorPosition/build)]
+              (dotimes [_n n]
+                (.logPut live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
+                         0 0 #(.getPositionAndIncrement wp))))
 
-          (.commit live-table-tx)
+            (.commit live-table-tx)
 
-          (let [leaves (.leaves (.compactLogs ^LiveHashTrie (.live-trie ^TestLiveTable live-table)))
-                leaf ^LiveHashTrie$Leaf (first leaves)]
+            (let [leaves (.leaves (.compactLogs ^LiveHashTrie (.live-trie ^TestLiveTable live-table)))
+                  leaf ^LiveHashTrie$Leaf (first leaves)]
 
-            (t/is (= 1 (count leaves)))
+              (t/is (= 1 (count leaves)))
 
-            (t/is
-             (uuid-equal-to-path?
-              uuid
-              (.path leaf)))
+              (t/is
+               (uuid-equal-to-path?
+                uuid
+                (.path leaf)))
 
-            (t/is (= (reverse (range n))
-                     (->> leaf
-                          (.data)
-                          (vec)))))
+              (t/is (= (reverse (range n))
+                       (->> leaf
+                            (.data)
+                            (vec)))))
 
-          @(.finishChunk live-table 0)
+            @(.finishChunk live-table 0)
 
-          (tu/with-allocator
-            #(tj/check-json
-              (.toPath (io/as-file (io/resource "xtdb/live-table-test/max-depth-trie-s")))
-              path)))))
+            (tu/with-allocator
+              #(tj/check-json
+                (.toPath (io/as-file (io/resource "xtdb/live-table-test/max-depth-trie-s")))
+                path))))))
 
     (let [uuid #uuid "7fffffff-ffff-ffff-4fff-ffffffffffff"
           n 50000]
       (tu/with-tmp-dirs #{path}
         (with-open [obj-store (obj-store-test/fs path)
                     allocator (RootAllocator.)
-                    live-table ^ILiveTable (live-index/->live-table
-                                            allocator obj-store "foo")
-                    live-table-tx (.startTx
-                                   live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))]
+                    live-table (live-index/->live-table allocator obj-store "foo")]
+          (let [live-table-tx (.startTx live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")) false)]
 
-          (let [wp (IVectorPosition/build)]
-            (dotimes [_n n]
-              (.logPut
-               live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
-               0 0 #(.getPositionAndIncrement wp))))
+            (let [wp (IVectorPosition/build)]
+              (dotimes [_n n]
+                (.logPut live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid)) 0 0 #(.getPositionAndIncrement wp))))
 
-          (.commit live-table-tx)
+            (.commit live-table-tx)
 
-          (let [leaves (.leaves (.compactLogs ^LiveHashTrie (.live-trie ^TestLiveTable live-table)))
-                leaf ^LiveHashTrie$Leaf (first leaves)]
+            (let [leaves (.leaves (.compactLogs ^LiveHashTrie (.live-trie ^TestLiveTable live-table)))
+                  leaf ^LiveHashTrie$Leaf (first leaves)]
 
-            (t/is (= 1 (count leaves)))
+              (t/is (= 1 (count leaves)))
 
-            (t/is
-             (uuid-equal-to-path?
-              uuid
-              (.path leaf)))
+              (t/is
+               (uuid-equal-to-path?
+                uuid
+                (.path leaf)))
 
-            (t/is (= (reverse (range n))
-                     (->> leaf
-                          (.data)
-                          (vec)))))
+              (t/is (= (reverse (range n))
+                       (->> leaf
+                            (.data)
+                            (vec)))))
 
-          @(.finishChunk live-table 0)
+            @(.finishChunk live-table 0)
 
-          (tu/with-allocator
-            #(tj/check-json
-              (.toPath (io/as-file (io/resource "xtdb/live-table-test/max-depth-trie-l")))
-              path)))))))
+            (tu/with-allocator
+              #(tj/check-json
+                (.toPath (io/as-file (io/resource "xtdb/live-table-test/max-depth-trie-l")))
+                path))))))))
 
 (defn live-table-wm->data [^ILiveTableWatermark live-table-wm]
   (let [live-rel-data (vr/rel->rows (.liveRelation live-table-wm))
@@ -128,63 +118,58 @@
   (let [uuids [#uuid "7fffffff-ffff-ffff-4fff-ffffffffffff"]]
     (with-open [obj-store (obj-store-test/in-memory)
                 allocator (RootAllocator.)
-                live-table ^ILiveTable (live-index/->live-table
-                                         allocator obj-store "foo")
-                live-table-tx (.startTx
-                                live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))]
+                live-table (live-index/->live-table allocator obj-store "foo")]
+      (let [live-table-tx (.startTx live-table (xtp/->TransactionInstant 0 (.toInstant #inst "2000")) false)]
 
-      (let [wp (IVectorPosition/build)]
-        (doseq [uuid uuids]
-                (.logPut
-                  live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
-                  0 0 #(.getPositionAndIncrement wp))))
+        (let [wp (IVectorPosition/build)]
+          (doseq [uuid uuids]
+            (.logPut
+             live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
+             0 0 #(.getPositionAndIncrement wp))))
 
-      (.commit live-table-tx)
+        (.commit live-table-tx)
 
-      (with-open [live-table-wm ^AutoCloseable (.openWatermark live-table true)]
+        (with-open [live-table-wm (.openWatermark live-table true)]
+          (let [live-table-before (live-table-wm->data live-table-wm)]
 
-        (let [live-table-before (live-table-wm->data live-table-wm)]
+            @(.finishChunk live-table 0)
+            (.close live-table)
 
-          @(.finishChunk live-table 0)
-          (.close live-table)
+            (let [live-table-after (live-table-wm->data live-table-wm)]
 
-          (let [live-table-after (live-table-wm->data live-table-wm)]
+              (t/is (= (:live-trie-iids live-table-before)
+                       (:live-trie-iids live-table-after)
+                       uuids))
 
-            (t/is (= (:live-trie-iids live-table-before)
-                     (:live-trie-iids live-table-after)
-                     uuids))
-
-            (t/is (= live-table-before live-table-after))))))))
+              (t/is (= live-table-before live-table-after)))))))))
 
 (deftest test-live-index-watermarks-are-immutable
   (let [uuids [#uuid "7fffffff-ffff-ffff-4fff-ffffffffffff"]
         table-name "foo"]
     (with-open [obj-store (obj-store-test/in-memory)
                 allocator (RootAllocator.)
-                live-index ^ILiveIndex (live-index/->LiveIndex allocator obj-store (HashMap.) 64 1024)
-                live-index-tx (.startTx
-                                live-index (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))
-                live-table-tx ^ILiveTableTx (.liveTable live-index-tx table-name)]
+                ^ILiveIndex live-index (live-index/->LiveIndex allocator obj-store (HashMap.) 64 1024)]
+      (let [live-index-tx (.startTx live-index (xtp/->TransactionInstant 0 (.toInstant #inst "2000")))
+            live-table-tx (.liveTable live-index-tx table-name)]
 
-      (let [wp (IVectorPosition/build)]
-        (doseq [uuid uuids]
-                (.logPut
-                  live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
-                  0 0 #(.getPositionAndIncrement wp))))
+        (let [wp (IVectorPosition/build)]
+          (doseq [uuid uuids]
+            (.logPut
+             live-table-tx (ByteBuffer/wrap (util/uuid->bytes uuid))
+             0 0 #(.getPositionAndIncrement wp))))
 
-      (.commit live-index-tx)
+        (.commit live-index-tx)
 
-      (with-open [live-index-wm ^AutoCloseable (.openWatermark live-index)]
+        (with-open [live-index-wm (.openWatermark live-index)]
+          (let [live-table-before (live-table-wm->data (.liveTable live-index-wm table-name))]
 
-        (let [live-table-before (live-table-wm->data (.liveTable ^ILiveIndexWatermark live-index-wm table-name))]
+            (.finishChunk live-index 0)
+            (.close live-index)
 
-          (.finishChunk live-index 0)
-          (.close live-index)
+            (let [live-table-after (live-table-wm->data (.liveTable live-index-wm table-name))]
 
-          (let [live-table-after (live-table-wm->data (.liveTable ^ILiveIndexWatermark live-index-wm table-name))]
+              (t/is (= (:live-trie-iids live-table-before)
+                       (:live-trie-iids live-table-after)
+                       uuids))
 
-            (t/is (= (:live-trie-iids live-table-before)
-                     (:live-trie-iids live-table-after)
-                     uuids))
-
-            (t/is (= live-table-before live-table-after))))))))
+              (t/is (= live-table-before live-table-after)))))))))

